File: Python3/Exclusive Time of Functions.py
# 84ms, 27.83%
class Solution:
    def exclusiveTime(self, n, logs):
        """
        :type n: int
        :type logs: List[str]
        :rtype: List[int]
        """

        times, stack = [0] * n, []
        last_time = 0

        for log in logs:
            cur_id, state, cur_time = log.split(':')
            cur_id, cur_time = int(cur_id), int(cur_time)
            if state == 's':
                if stack: times[stack[-1]] += cur_time - last_time
                stack.append(cur_id)
                last_time = cur_time
            else:
                times[stack.pop()] += cur_time - last_time + 1
                last_time = cur_time + 1
        return times

File: Cpp/Exclusive Time of Functions.cpp
// 12ms, 100.0%
static int x = []() {
    ios_base::sync_with_stdio(false);
    cin.tie(nullptr);
    return 0;
}();

class Solution
{
  public:
    vector<int> exclusiveTime(int n, const vector<string> &logs)
    {
        vector<int> times;
        times.resize(n);

        stack<int> st;
        st.push(getFunctionId(logs.front()));
        int lastTime = getTimeStamp(logs.front());
        for (int i = 1, num = logs.size(); i < num; ++i) {
            auto &x = logs[i];
            int curFunc = getFunctionId(x), curTime = getTimeStamp(x);
            if (getFunctionState(x)) {  // start
                times[st.top()] += curTime - lastTime;
                st.push(curFunc);
                lastTime = curTime;
            } else {  // end
                times[curFunc] += curTime - lastTime + 1;
                st.pop();
                lastTime = curTime + 1;
            }
        }
        return times;
    }

  private:
    int getFunctionId(const string &s)
    {
        int x = s.find(':');
        return stoi(s.substr(0, x));
    }

    bool getFunctionState(const string &s)
    {
        int x = s.find(':');
        return (s[x + 1] == 's' ? true : false);
    }

    int getTimeStamp(const string &s)
    {
        int x = s.rfind(':');
        return stoi(s.substr(x + 1));
    }
};
